https://github.com/torvalds/linux/commit/b9a85e5eec126d6ae6c362f94b447c223e8fe6e4

RDMA/usnic: Silence uninitialized symbol smatch warnings

The patch 1da177e: "Linux-2.6.12-rc2" from Apr 16, 2005
(linux-next), leads to the following Smatch static checker warning:

        drivers/infiniband/hw/mthca/mthca_cmd.c:644 mthca_SYS_EN()
        error: uninitialized symbol 'out'.

drivers/infiniband/hw/mthca/mthca_cmd.c
    636 int mthca_SYS_EN(struct mthca_dev *dev)
    637 {
    638         u64 out;
    639         int ret;
    640
    641         ret = mthca_cmd_imm(dev, 0, &out, 0, 0, CMD_SYS_EN, CMD_TIME_CLASS_D);

We pass out here and it gets used without being initialized.

        err = mthca_cmd_post(dev, in_param,
                             out_param ? *out_param : 0,
                                         ^^^^^^^^^^
                             in_modifier, op_modifier,
                             op, context->token, 1);

It's the same in mthca_cmd_wait() and mthca_cmd_poll().

Fixes: 1da177e ("Linux-2.6.12-rc2")
Reported-by: Dan Carpenter <dan.carpenter@linaro.org>
Closes: https://lore.kernel.org/all/533bc3df-8078-4397-b93d-d1f6cec9b636@moroto.mountain
Link: https://lore.kernel.org/r/c559cb7113158c02d75401ac162652072ef1b5f0.1699867650.git.leon@kernel.org
Signed-off-by: Leon Romanovsky <leonro@nvidia.com>

diff -Naurp ib_mthca-1.0.20080404.orig/mthca_cmd.c ib_mthca-1.0.20080404/mthca_cmd.c
--- ib_mthca-1.0.20080404.orig/mthca_cmd.c	2024-04-10 09:08:06.000000000 -0400
+++ ib_mthca-1.0.20080404/mthca_cmd.c	2024-09-09 16:09:03.689936851 -0400
@@ -635,7 +635,7 @@ void mthca_free_mailbox(struct mthca_dev
 
 int mthca_SYS_EN(struct mthca_dev *dev)
 {
-	u64 out;
+	u64 out = 0;
 	int ret;
 
 	ret = mthca_cmd_imm(dev, 0, &out, 0, 0, CMD_SYS_EN, CMD_TIME_CLASS_D);
@@ -1955,7 +1955,7 @@ int mthca_WRITE_MGM(struct mthca_dev *de
 int mthca_MGID_HASH(struct mthca_dev *dev, struct mthca_mailbox *mailbox,
 		    u16 *hash)
 {
-	u64 imm;
+	u64 imm = 0;
 	int err;
 
 	err = mthca_cmd_imm(dev, mailbox->dma, &imm, 0, 0, CMD_MGID_HASH,
diff -Naurp ib_mthca-1.0.20080404.orig/mthca_main.c ib_mthca-1.0.20080404/mthca_main.c
--- ib_mthca-1.0.20080404.orig/mthca_main.c	2024-04-10 09:08:06.000000000 -0400
+++ ib_mthca-1.0.20080404/mthca_main.c	2024-09-09 16:09:36.689464702 -0400
@@ -382,7 +382,7 @@ static int mthca_init_icm(struct mthca_d
 			  struct mthca_init_hca_param *init_hca,
 			  u64 icm_size)
 {
-	u64 aux_pages;
+	u64 aux_pages = 0;
 	int err;
 
 	err = mthca_SET_ICM_SIZE(mdev, icm_size, &aux_pages);
